HotReload (extensions\vertx-http\deployment) tests fails on Windows with Java 11

HotReload (extensions\vertx-http\deployment) tests fails on **Windows with Java 11** (Java version: 11.0.5, vendor: Oracle Corporation)

`Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 4: file:/C:/Users/Rostislav/.m2/repository/org/apache/maven/surefire/surefire-booter/2.22.1/surefire-booter-2.22.1.jar`

```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   HotReloadWithFilterTest.testFilterChange:34 1 expectation failed.
Expected status code <200> but was <500>.

[ERROR] Errors:
[ERROR]   HotReloadWithFilterTest.testAddFilter » Runtime java.lang.RuntimeException: ja...
[ERROR]   HotReloadWithRouteTest.testAddBean » Runtime java.lang.RuntimeException: java....
[ERROR]   HotReloadWithRouteTest.testRouteChange » Runtime java.lang.RuntimeException: j...
[INFO]
[ERROR] Tests run: 41, Failures: 1, Errors: 3, Skipped: 0
```

Here is [test-log.txt](https://github.com/quarkusio/quarkus/files/3987194/test-log.txt) from `mvn clean verify -fae -pl extensions\vertx-http\deployment\` command execution.

```
$ mvn -version
Apache Maven 3.5.3 (3383c37e1f9e9b3bc3df5050c29c8aff9f295297; 2018-02-24T20:49:05+01:00)
Maven home: c:\Users\Rostislav\Downloads\apache-maven-3.5.3\bin\..
Java version: 11.0.5, vendor: Oracle Corporation
Java home: c:\Program Files\Java\jdk-11.0.5
Default locale: cs_CZ, platform encoding: Cp1250
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

@stuartwdouglas / @gsmet  fyi

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
@@ -4,6 +4,8 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.UncheckedIOException;
+import java.net.URL;
+import java.net.URLDecoder;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.FileVisitResult;
 import java.nio.file.FileVisitor;
@@ -15,10 +17,13 @@ import java.nio.file.attribute.FileTime;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Enumeration;
 import java.util.List;
 import java.util.ServiceLoader;
 import java.util.function.Function;
 import java.util.function.Supplier;
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
 import java.util.stream.Stream;
 
 import org.jboss.shrinkwrap.api.exporter.ExplodedExporter;
@@ -45,13 +50,12 @@ import io.quarkus.test.common.http.TestHTTPResourceManager;
  * testing their extension functionality in dev mode. Unlike {@link QuarkusUnitTest} this will test against
  * a clean deployment for each test method. This is nessesary to prevent undefined behaviour by making sure the
  * deployment starts in a clean state for each test.
- *
- *
+ * <p>
+ * <p>
  * NOTE: These tests do not run with {@link io.quarkus.runtime.LaunchMode#TEST} but rather with
  * {@link io.quarkus.runtime.LaunchMode#DEVELOPMENT}. This is necessary to ensure dev mode is tested correctly.
- *
+ * <p>
  * A side effect of this is that the tests will run on port 8080 by default instead of port 8081.
- *
  */
 public class QuarkusDevModeTest
         implements BeforeEachCallback, AfterEachCallback, TestInstanceFactory {
@@ -275,9 +279,61 @@ public class QuarkusDevModeTest
                             Collections.singleton(deploymentSourcePath.toAbsolutePath().toString()),
                             classes.toAbsolutePath().toString(), deploymentResourcePath.toAbsolutePath().toString()));
 
+            setDevModeRunnerJarFile(context);
             return context;
         } catch (Exception e) {
             throw new RuntimeException("Unable to create the archive", e);
+        }
+    }
+
+    private static void setDevModeRunnerJarFile(final DevModeContext context) {
+        try {
+            /*
+             * See https://github.com/quarkusio/quarkus/issues/6280
+             * Maven surefire plugin launches the (forked) JVM for tests using a "surefirebooter" jar file.
+             * This jar file's name starts with the prefix "surefirebooter" and ends with the extension ".jar".
+             * The jar is launched using "java -jar .../surefirebooter*.jar ..." semantics. This jar has a
+             * MANIFEST which contains "Class-Path" entries. These entries trigger a bug in the JDK code
+             * https://bugs.openjdk.java.net/browse/JDK-8232170 which causes hot deployment related logic in Quarkus
+             * to fail in dev mode.
+             * The goal in this next section is to narrow down to this specific surefirebooter*.jar which was used to launch
+             * the tests and mark it as the "dev mode runner jar" (through DevModeContext#setDevModeRunnerJarFile),
+             * so that programmatic compilation of code (during hot deployment) doesn't run into issues noted in
+             * https://bugs.openjdk.java.net/browse/JDK-8232170.
+             * In reality the surefirebooter*.jar isn't really a "dev mode runner jar" (i.e. it's not the -dev.jar that
+             * Quarkus generates), but it's fine to mark it as such to get past this issue. This is more of a workaround
+             * on top of another workaround. In the medium/long term the actual JDK issue fix will make its way into
+             * almost all prominently used Java versions.
+             */
+            final Enumeration<URL> manifests = QuarkusDevModeTest.class.getClassLoader().getResources("META-INF/MANIFEST.MF");
+            while (manifests.hasMoreElements()) {
+                final URL url = manifests.nextElement();
+                // don't open streams to manifest entries unless it resembles to the one
+                // we are interested in
+                if (!url.getPath().contains("surefirebooter")) {
+                    continue;
+                }
+                try (final InputStream is = url.openStream()) {
+                    final Manifest manifest = new Manifest(is);
+                    final String mainClass = manifest.getMainAttributes().getValue(Attributes.Name.MAIN_CLASS);
+                    // additional check to make sure we are probing the right jar
+                    if ("org.apache.maven.surefire.booter.ForkedBooter".equals(mainClass)) {
+                        final String manifestFilePath = url.getPath();
+                        if (manifestFilePath.startsWith("file:")) {
+                            // manifest file path will be of the form jar:file:....!META-INF/MANIFEST.MF
+                            final String jarFilePath = manifestFilePath.substring(5, manifestFilePath.lastIndexOf('!'));
+                            final File surefirebooterJar = new File(
+                                    URLDecoder.decode(jarFilePath, StandardCharsets.UTF_8.name()));
+                            context.setDevModeRunnerJarFile(surefirebooterJar);
+                        }
+                        break;
+                    }
+
+                }
+            }
+        } catch (Throwable t) {
+            // ignore and move on
+            return;
         }
     }
 
@@ -303,7 +359,7 @@ public class QuarkusDevModeTest
 
     /**
      * Adds the source file that corresponds to the given class to the deployment
-     * 
+     *
      * @param sourceFile
      */
     public void addSourceFile(Class<?> sourceFile) {

